File: src/stores/lmStudio.ts
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

/**
 * LM Studio 本地模型管理模块
 *
 * 负责连接 LM Studio 本地服务器并管理其模型 (列表/下载/加载/卸载)。
 * 与 Ollama 不同，LM Studio 是闭源 GUI 桌面应用，没有公开的无人值守
 * 安装方式，因此本模块只连接一个已经在运行的 LM Studio 服务，不负责
 * 下载/安装 LM Studio 本身 -- 用户需要自行从官网下载安装并在其设置里
 * 启动本地服务器 (Developer 标签页 -> Start Server)。
 */

import { ref, computed } from "vue";
import { defineStore } from "pinia";
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { useSettingsStore } from "./settings";

/**
 * LM Studio 模型信息
 * 与后端 LMStudioModelInfo 结构对应
 */
export interface LMStudioModelInfo {
  id: string;
  modelType: string;
  publisher: string | null;
  arch: string | null;
  compatibilityType: string | null;
  quantization: string | null;
  /** "loaded" | "not-loaded" */
  state: string;
  maxContextLength: number | null;
}

/**
 * 模型下载进度事件
 */
export interface LMStudioDownloadProgress {
  modelId: string;
  /** "downloading" | "paused" | "completed" | "failed" */
  status: string;
  downloadedBytes: number | null;
  totalSizeBytes: number | null;
}

export const useLMStudioStore = defineStore(
  "lmStudio",
  () => {
    // ============ 响应式状态 ============

    /** LM Studio 服务器地址 (不含 /v1，由各命令自行拼接路径) */
    const baseUrl = ref("http://localhost:1234");

    /** 可选的 API Key (仅当用户在 LM Studio 中开启了 "Require API key" 时需要) */
    const apiKey = ref("");

    /** LM Studio 服务是否在线 */
    const isOnline = ref(false);

    /** 模型列表 (已下载 / 已加载) */
    const models = ref<LMStudioModelInfo[]>([]);

    /** 是否正在加载模型列表 */
    const isLoadingModels = ref(false);

    /** 是否正在下载模型 */
    const isPulling = ref(false);

    /** 当前正在下载的模型 ID */
    const pullingModelId = ref("");

    /** 下载进度信息 */
    const downloadProgress = ref<LMStudioDownloadProgress | null>(null);

    /** 下载进度监听器取消函数 */
    let unlistenDownload: UnlistenFn | null = null;

    // ============ 计算属性 ============

    /** 下载进度百分比 */
    const downloadPercent = computed(() => {
      if (!downloadProgress.value) return null;
      const { downloadedBytes, totalSizeBytes } = downloadProgress.value;
      if (downloadedBytes == null || totalSizeBytes == null || totalSizeBytes === 0) {
        return null;
      }
      return Math.round((downloadedBytes / totalSizeBytes) * 100);
    });

    /** 当前已加载到内存中的模型 */
    const loadedModels = computed(() => models.value.filter((m) => m.state === "loaded"));

    // ============ 方法函数 ============

    /**
     * 检查 LM Studio 服务状态
     */
    const checkStatus = async () => {
      try {
        isOnline.value = await invoke<boolean>("check_lmstudio_status", {
          baseUrl: baseUrl.value,
        });
      } catch (error) {
        console.error("Failed to check LM Studio status:", error);
        isOnline.value = false;
      }
    };

    /**
     * 加载模型列表
     */
    const loadModels = async () => {
      if (!isOnline.value) {
        await checkStatus();
        if (!isOnline.value) return;
      }

      isLoadingModels.value = true;
      try {
        models.value = await invoke<LMStudioModelInfo[]>("list_lmstudio_models", {
          baseUrl: baseUrl.value,
          apiKey: apiKey.value || undefined,
        });
      } catch (error) {
        console.error("Failed to load LM Studio models:", error);
        models.value = [];
      } finally {
        isLoadingModels.value = false;
      }
    };

    /**
     * 设置 LM Studio 服务地址
     */
    const setBaseUrl = async (url: string) => {
      baseUrl.value = url;
      await checkStatus();
      if (isOnline.value) {
        await loadModels();
      }
    };

    /**
     * 设置下载进度监听器
     */
    const setupDownloadListener = async () => {
      if (unlistenDownload) {
        unlistenDownload();
      }

      unlistenDownload = await listen<LMStudioDownloadProgress>(
        "lmstudio-download-progress",
        (event) => {
          downloadProgress.value = event.payload;
        }
      );
    };

    /**
     * 下载模型
     * @param modelId 模型标识 (LM Studio 模型目录名或 HuggingFace 引用)
     */
    const pullModel = async (modelId: string) => {
      if (isPulling.value) return;

      isPulling.value = true;
      pullingModelId.value = modelId;
      downloadProgress.value = null;

      try {
        await setupDownloadListener();

        await invoke("pull_lmstudio_model", {
          modelId,
          baseUrl: baseUrl.value,
          apiKey: apiKey.value || undefined,
        });

        await loadModels();
      } catch (error) {
        console.error("Failed to pull LM Studio model:", error);
        throw error;
      } finally {
        isPulling.value = false;
        pullingModelId.value = "";
        downloadProgress.value = null;
        if (unlistenDownload) {
          unlistenDownload();
          unlistenDownload = null;
        }
      }
    };

    /**
     * 加载模型到内存
     */
    const loadModel = async (modelId: string) => {
      try {
        await invoke("load_lmstudio_model", {
          modelId,
          baseUrl: baseUrl.value,
          apiKey: apiKey.value || undefined,
        });
        await loadModels();
      } catch (error) {
        console.error("Failed to load LM Studio model:", error);
        throw error;
      }
    };

    /**
     * 从内存卸载模型
     */
    const unloadModel = async (modelId: string) => {
      try {
        await invoke("unload_lmstudio_model", {
          modelId,
          baseUrl: baseUrl.value,
          apiKey: apiKey.value || undefined,
        });
        await loadModels();
      } catch (error) {
        console.error("Failed to unload LM Studio model:", error);
        throw error;
      }
    };

    /**
     * 格式化文件大小
     */
    const formatSize = (bytes: number): string => {
      if (bytes === 0) return "0 B";
      const units = ["B", "KB", "MB", "GB", "TB"];
      const i = Math.floor(Math.log(bytes) / Math.log(1024));
      const size = bytes / Math.pow(1024, i);
      return `${size.toFixed(i > 0 ? 1 : 0)} ${units[i]}`;
    };

    /**
     * 一键部署：确保模型已加载 -> 创建聊天 API 配置
     * 复用 "local" provider，因为 LM Studio 同样暴露了无需鉴权的
     * OpenAI 兼容 /v1/chat/completions 接口，和 Ollama 走的是同一套
     * 聊天请求代码路径
     */
    const oneClickDeploy = async (modelId: string) => {
      if (!isOnline.value) {
        throw new Error("LM Studio 服务未连接");
      }

      const model = models.value.find((m) => m.id === modelId);
      if (model && model.state !== "loaded") {
        await loadModel(modelId);
      }

      const chatBaseUrl = `${baseUrl.value.replace(/\/$/, "")}/v1`;

      const settings = useSettingsStore();
      const existingConfig = settings.apiConfigs.find(
        (c) => c.provider === "local" && c.model === modelId && c.baseUrl === chatBaseUrl
      );

      if (!existingConfig) {
        const configName = `lmstudio-${modelId}`;
        settings.createApiConfig(
          configName,
          "local",
          modelId,
          "", // LM Studio 默认无需 API Key
          chatBaseUrl
        );
      }

      return modelId;
    };

    // ============ 返回公共接口 ============
    return {
      // 状态
      baseUrl,
      apiKey,
      isOnline,
      models,
      isLoadingModels,
      isPulling,
      pullingModelId,
      downloadProgress,

      // 计算属性
      downloadPercent,
      loadedModels,

      // 方法
      checkStatus,
      loadModels,
      setBaseUrl,
      pullModel,
      loadModel,
      unloadModel,
      formatSize,
      oneClickDeploy,
    };
  },
  {
    persist: {
      key: "baiyu-aispace-lmstudio",
      paths: ["baseUrl", "apiKey"],
    },
  }
);

File: src-tauri/src/commands/lmstudio.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

//! LM Studio 本地服务器集成
//!
//! LM Studio 会跑自己的本地推理服务器（默认是 `http://localhost:1234`），
//! 同时提供兼容 OpenAI 的接口面（`/v1/chat/completions`、`/v1/models`）和
//! 一套原生的 v1 管理 API（`/api/v1/models/download`、`/load`、`/unload`），
//! 本模块用后者做模型管理；此外还有一个信息更丰富的 v0 列表接口
//! （`/api/v0/models`），会报告每个模型的发布者/架构/量化方式/加载状态。
//! 与 Ollama 不同，LM Studio 是闭源的 GUI 桌面应用，没有官方文档记载的
//! 无头/静默安装方式，所以本模块自始至终只跟一个已经在运行的服务器通信——
//! 它从不尝试安装或启动 LM Studio 本身。

use serde::{Deserialize, Serialize};
use std::time::Duration;
use tauri::{AppHandle, Emitter};

use super::local_model::friendly_err;

// ============ 类型定义 ============

/// LM Studio `/api/v0/models` 接口返回的模型信息。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LMStudioModelInfo {
    pub id: String,
    pub model_type: String,
    pub publisher: Option<String>,
    pub arch: Option<String>,
    pub compatibility_type: Option<String>,
    pub quantization: Option<String>,
    /// "loaded" | "not-loaded"
    pub state: String,
    pub max_context_length: Option<u64>,
}

#[derive(Debug, Deserialize)]
struct LMStudioModelsResponse {
    data: Vec<LMStudioModelEntry>,
}

#[derive(Debug, Deserialize)]
struct LMStudioModelEntry {
    id: String,
    #[serde(rename = "type")]
    model_type: Option<String>,
    publisher: Option<String>,
    arch: Option<String>,
    compatibility_type: Option<String>,
    quantization: Option<String>,
    state: Option<String>,
    max_context_length: Option<u64>,
}

/// 下发给前端的下载进度事件。
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LMStudioDownloadProgress {
    pub model_id: String,
    /// "downloading" | "paused" | "completed" | "failed"
    pub status: String,
    pub downloaded_bytes: Option<u64>,
    pub total_size_bytes: Option<u64>,
}

// ============ 辅助函数 ============

fn create_lmstudio_client() -> reqwest::Result<reqwest::Client> {
    reqwest::Client::builder()
        .timeout(Duration::from_secs(300))
        .connect_timeout(Duration::from_secs(10))
        // 这个客户端访问的永远是用户自己已经在运行的 LM Studio 服务器，
        // 绝不是远程主机 -- 如果按 reqwest 默认行为走系统代理，在用户为
        // 其他服务商开着全局代理模式时，会平白多绕几秒钟。
        .no_proxy()
        .build()
}

fn build_lmstudio_url(base_url: &str, endpoint: &str) -> String {
    let base = base_url.trim_end_matches('/');
    format!("{}{}", base, endpoint)
}

/// LM Studio 的本地服务器默认没有鉴权；只有用户在其设置里显式开启了
/// "Require API key" 才需要 API token，所以只有真正配置了密钥时才会加这个请求头。
fn auth_headers(api_key: &Option<String>) -> reqwest::header::HeaderMap {
    let mut headers = reqwest::header::HeaderMap::new();
    if let Some(key) = api_key {
        if !key.is_empty() {
            if let Ok(value) = format!("Bearer {}", key).parse() {
                headers.insert(reqwest::header::AUTHORIZATION, value);
            }
        }
    }
    headers
}

// ============ Tauri 命令 ============

/// 检查 `base_url` 处的 LM Studio 服务器是否可达。
#[tauri::command]
pub async fn check_lmstudio_status(base_url: String) -> Result<bool, String> {
    let client = create_lmstudio_client().map_err(|e| friendly_err("创建网络连接失败，请重启应用后重试", e))?;
    let url = build_lmstudio_url(&base_url, "/v1/models");

    match client.get(&url).send().await {
        Ok(response) => Ok(response.status().is_success()),
        Err(e) => {
            log::debug!("LM Studio status check failed: {}", e);
            Ok(false)
        }
    }
}

/// 列出 LM Studio 已知的模型（已下载和/或当前已加载的）。
#[tauri::command]
pub async fn list_lmstudio_models(
    base_url: String,
    api_key: Option<String>,
) -> Result<Vec<LMStudioModelInfo>, String> {
    let client = create_lmstudio_client().map_err(|e| friendly_err("创建网络连接失败，请重启应用后重试", e))?;
    let url = build_lmstudio_url(&base_url, "/api/v0/models");

    let response = client
        .get(&url)
        .headers(auth_headers(&api_key))
        .send()
        .await
        .map_err(|e| friendly_err("无法连接到 LM Studio，请确认 LM Studio 服务器已启动", e))?;

    if !response.status().is_success() {
        return Err(friendly_err("LM Studio 返回异常状态，请检查服务是否正常运行", response.status()));
    }

    let body: LMStudioModelsResponse = response
        .json()
        .await
        .map_err(|e| friendly_err("解析 LM Studio 返回的数据失败，请确认版本是否兼容", e))?;

    Ok(body
        .data
        .into_iter()
        .map(|m| LMStudioModelInfo {
            id: m.id,
            model_type: m.model_type.unwrap_or_default(),
            publisher: m.publisher,
            arch: m.arch,
            compatibility_type: m.compatibility_type,
            quantization: m.quantization,
            state: m.state.unwrap_or_else(|| "unknown".to_string()),
            max_context_length: m.max_context_length,
        })
        .collect())
}

/// 通过目录标识符（如 `"qwen2.5-7b-instruct"`）或 Hugging Face 引用下载模型，
/// 在轮询返回的任务直到完成或失败的过程中，持续下发 `lmstudio-download-progress` 事件。
#[tauri::command]
pub async fn pull_lmstudio_model(
    model_id: String,
    base_url: String,
    api_key: Option<String>,
    app_handle: AppHandle,
) -> Result<(), String> {
    let client = create_lmstudio_client().map_err(|e| friendly_err("创建网络连接失败，请重启应用后重试", e))?;
    let download_url = build_lmstudio_url(&base_url, "/api/v1/models/download");

    log::info!("Downloading LM Studio model: {}", model_id);

    let response = client
        .post(&download_url)
        .headers(auth_headers(&api_key))
        .json(&serde_json::json!({ "model": model_id }))
        .send()
        .await
        .map_err(|e| friendly_err("无法开始下载模型，请检查网络连接和 LM Studio 服务状态", e))?;

    if !response.status().is_success() {
        let error_text = response.text().await.unwrap_or_default();
        return Err(friendly_err("模型下载失败，请确认模型标识是否正确", error_text));
    }

    let job: serde_json::Value = response
        .json()
        .await
        .map_err(|e| friendly_err("解析下载任务响应失败", e))?;

    // 模型可能已经在磁盘上了，这种情况下没有任务需要轮询。
    let initial_status = job["status"].as_str().unwrap_or("").to_string();
    if initial_status == "already_downloaded" || initial_status == "completed" {
        let _ = app_handle.emit("lmstudio-download-progress", LMStudioDownloadProgress {
            model_id,
            status: "completed".to_string(),
            downloaded_bytes: job["total_size_bytes"].as_u64(),
            total_size_bytes: job["total_size_bytes"].as_u64(),
        });
        return Ok(());
    }

    let job_id = job["job_id"]
        .as_str()
        .ok_or("下载任务响应格式异常，请重试")?
        .to_string();

    let status_url = build_lmstudio_url(
        &base_url,
        &format!("/api/v1/models/download/status/{}", job_id),
    );

    loop {
        tokio::time::sleep(Duration::from_millis(500)).await;

        let status_response = client
            .get(&status_url)
            .headers(auth_headers(&api_key))
            .send()
            .await
            .map_err(|e| friendly_err("查询下载进度失败，请检查网络连接", e))?;

        if !status_response.status().is_success() {
            return Err(friendly_err(
                "查询下载进度失败，请检查 LM Studio 服务是否正常运行",
                status_response.status(),
            ));
        }

        let status_json: serde_json::Value = status_response
            .json()
            .await
            .map_err(|e| friendly_err("解析下载进度失败", e))?;

        let status = status_json["status"].as_str().unwrap_or("").to_string();

        let _ = app_handle.emit("lmstudio-download-progress", LMStudioDownloadProgress {
            model_id: model_id.clone(),
            status: status.clone(),
            downloaded_bytes: status_json["downloaded_bytes"].as_u64(),
            total_size_bytes: status_json["total_size_bytes"].as_u64(),
        });

        match status.as_str() {
            "completed" => {
                log::info!("LM Studio model download completed: {}", model_id);
                return Ok(());
            }
            "failed" => {
                return Err(format!("模型 {} 下载失败，请重试或更换模型", model_id));
            }
            _ => continue, // "downloading" / "paused" -- 继续轮询
        }
    }
}

/// 把已下载的模型加载进内存，使其可用于推理。
#[tauri::command]
pub async fn load_lmstudio_model(
    model_id: String,
    base_url: String,
    api_key: Option<String>,
) -> Result<(), String> {
    let client = create_lmstudio_client().map_err(|e| friendly_err("创建网络连接失败，请重启应用后重试", e))?;
    let url = build_lmstudio_url(&base_url, "/api/v1/models/load");

    let response = client
        .post(&url)
        .headers(auth_headers(&api_key))
        .json(&serde_json::json!({ "model": model_id }))
        .send()
        .await
        .map_err(|e| friendly_err("加载模型失败，请确认 LM Studio 服务正在运行", e))?;

    if !response.status().is_success() {
        let error_text = response.text().await.unwrap_or_default();
        return Err(friendly_err("加载模型失败，请检查模型是否已下载完整", error_text));
    }

    log::info!("LM Studio model loaded: {}", model_id);
    Ok(())
}

/// 卸载已加载的模型以释放内存。
#[tauri::command]
pub async fn unload_lmstudio_model(
    model_id: String,
    base_url: String,
    api_key: Option<String>,
) -> Result<(), String> {
    let client = create_lmstudio_client().map_err(|e| friendly_err("创建网络连接失败，请重启应用后重试", e))?;
    let url = build_lmstudio_url(&base_url, "/api/v1/models/unload");

    let response = client
        .post(&url)
        .headers(auth_headers(&api_key))
        .json(&serde_json::json!({ "model": model_id }))
        .send()
        .await
        .map_err(|e| friendly_err("卸载模型失败，请确认 LM Studio 服务正在运行", e))?;

    if !response.status().is_success() {
        let error_text = response.text().await.unwrap_or_default();
        return Err(friendly_err("卸载模型失败，请稍后重试", error_text));
    }

    log::info!("LM Studio model unloaded: {}", model_id);
    Ok(())
}
